File: study_os_service/domain/sessions.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import UTC, datetime
from typing import Literal


ProgressStatus = Literal["unread", "in_progress", "covered", "stale", "weak", "strong"]
SessionState = Literal["active", "finished"]
SessionOutcome = Literal["partial", "completed", "failed", "skipped", "abandoned"]
SkipReason = Literal[
    "lack_of_time",
    "fatigue",
    "wrong_material",
    "blocked_prerequisite",
    "too_difficult",
    "other",
]

_PROGRESS_STATUSES = {"unread", "in_progress", "covered", "stale", "weak", "strong"}
_SESSION_STATES = {"active", "finished"}
_SESSION_OUTCOMES = {"partial", "completed", "failed", "skipped", "abandoned"}
_SKIP_REASONS = {
    "lack_of_time",
    "fatigue",
    "wrong_material",
    "blocked_prerequisite",
    "too_difficult",
    "other",
}


def _positive_integer(value: int, label: str) -> None:
    if isinstance(value, bool) or not isinstance(value, int) or value < 1:
        raise ValueError(f"{label} must be a positive integer")


def _non_negative_integer(value: int, label: str) -> None:
    if isinstance(value, bool) or not isinstance(value, int) or value < 0:
        raise ValueError(f"{label} must be a non-negative integer")


def _datetime(value: datetime | None, label: str) -> datetime | None:
    if value is None:
        return None
    if value.tzinfo is None or value.utcoffset() is None:
        raise ValueError(f"{label} must be timezone-aware")
    return value.astimezone(UTC)


@dataclass(frozen=True, slots=True)
class ProgressState:
    id: int
    lesson_id: int
    material_id: int
    status: ProgressStatus
    cursor_page: int
    furthest_page: int
    completed_at: datetime | None
    last_seen_at: datetime | None
    confidence: float
    total_seconds: int
    session_count: int
    version: int

    def __post_init__(self) -> None:
        _positive_integer(self.id, "id")
        _positive_integer(self.lesson_id, "lesson id")
        _positive_integer(self.material_id, "material id")
        if self.status not in _PROGRESS_STATUSES:
            raise ValueError("invalid progress status")
        _positive_integer(self.cursor_page, "cursor page")
        _positive_integer(self.furthest_page, "furthest page")
        if self.furthest_page < self.cursor_page:
            raise ValueError("furthest page cannot precede cursor page")
        if isinstance(self.confidence, bool) or not isinstance(
            self.confidence, (int, float)
        ):
            raise ValueError("confidence must be numeric")
        if not 0 <= float(self.confidence) <= 1:
            raise ValueError("confidence must be between 0 and 1")
        _non_negative_integer(self.total_seconds, "total seconds")
        _non_negative_integer(self.session_count, "session count")
        _positive_integer(self.version, "version")
        object.__setattr__(
            self, "completed_at", _datetime(self.completed_at, "completed at")
        )
        object.__setattr__(
            self, "last_seen_at", _datetime(self.last_seen_at, "last seen at")
        )
        object.__setattr__(self, "confidence", float(self.confidence))


@dataclass(frozen=True, slots=True)
class StudySession:
    id: int
    idempotency_key: str
    target_slug: str
    lesson_id: int
    material_id: int
    state: SessionState
    started_at: datetime
    ended_at: datetime | None
    elapsed_seconds: int
    start_page: int
    end_page: int | None
    questions_done: int
    correct_count: int
    wrong_count: int
    doubt_count: int
    favorite_count: int
    outcome: SessionOutcome | None
    skip_reason: SkipReason | None
    notes: str
    version: int

    def __post_init__(self) -> None:
        _positive_integer(self.id, "id")
        key = self.idempotency_key.strip()
        target = self.target_slug.strip()
        if not key:
            raise ValueError("idempotency key is required")
        if not target:
            raise ValueError("target is required")
        object.__setattr__(self, "idempotency_key", key)
        object.__setattr__(self, "target_slug", target)
        _positive_integer(self.lesson_id, "lesson id")
        _positive_integer(self.material_id, "material id")
        if self.state not in _SESSION_STATES:
            raise ValueError("invalid session state")
        if self.outcome is not None and self.outcome not in _SESSION_OUTCOMES:
            raise ValueError("invalid session outcome")
        if self.skip_reason is not None and self.skip_reason not in _SKIP_REASONS:
            raise ValueError("invalid skip reason")

        started_at = _datetime(self.started_at, "started at")
        ended_at = _datetime(self.ended_at, "ended at")
        object.__setattr__(self, "started_at", started_at)
        object.__setattr__(self, "ended_at", ended_at)
        if ended_at is not None and ended_at < started_at:
            raise ValueError("ended at cannot precede started at")

        _non_negative_integer(self.elapsed_seconds, "elapsed seconds")
        _positive_integer(self.start_page, "start page")
        if self.end_page is not None:
            _positive_integer(self.end_page, "end page")
            if self.end_page < self.start_page:
                raise ValueError("end page cannot precede start page")
        for value, label in (
            (self.questions_done, "questions done"),
            (self.correct_count, "correct count"),
            (self.wrong_count, "wrong count"),
            (self.doubt_count, "doubt count"),
            (self.favorite_count, "favorite count"),
        ):
            _non_negative_integer(value, label)
        _positive_integer(self.version, "version")

        if self.state == "active":
            if self.outcome is not None or ended_at is not None:
                raise ValueError("active session cannot have an outcome or end time")
            if self.skip_reason is not None:
                raise ValueError("active session cannot have a skip reason")
        else:
            if self.outcome is None or ended_at is None:
                raise ValueError("finished session requires an outcome and end time")

        if self.outcome == "skipped":
            if self.skip_reason is None:
                raise ValueError("skipped session requires a skip reason")
        elif self.skip_reason is not None:
            raise ValueError("only skipped sessions may have a skip reason")
        if self.outcome in {"partial", "completed"} and self.end_page is None:
            raise ValueError(f"{self.outcome} session requires an end page")

File: src/study-os/api/sessions.ts
import { requestJson } from './client';

export type ProgressStatus = 'unread' | 'in_progress' | 'covered' | 'stale' | 'weak' | 'strong';
export type SessionState = 'active' | 'finished';
export type SessionOutcome = 'partial' | 'completed' | 'failed' | 'skipped' | 'abandoned';
export type SkipReason =
  | 'lack_of_time'
  | 'fatigue'
  | 'wrong_material'
  | 'blocked_prerequisite'
  | 'too_difficult'
  | 'other';

export interface ProgressState {
  id: number;
  lessonId: number;
  materialId: number;
  status: ProgressStatus;
  cursorPage: number;
  furthestPage: number;
  completedAt: string | null;
  lastSeenAt: string | null;
  confidence: number;
  totalSeconds: number;
  sessionCount: number;
  version: number;
}

export interface StudySession {
  id: number;
  idempotencyKey: string;
  targetSlug: string;
  lessonId: number;
  materialId: number;
  state: SessionState;
  startedAt: string;
  endedAt: string | null;
  elapsedSeconds: number;
  startPage: number;
  endPage: number | null;
  questionsDone: number;
  correctCount: number;
  wrongCount: number;
  doubtCount: number;
  favoriteCount: number;
  outcome: SessionOutcome | null;
  skipReason: SkipReason | null;
  notes: string;
  version: number;
}

export interface MaterialInspection {
  materialId: number;
  pageCount: number;
  pageOffset: number;
}

export interface ReadingRateSummary {
  materialId: number;
  pagesPerHour: number;
  sampleCount: number;
  totalSeconds: number;
  source: 'default' | 'observed';
}

export interface ReadingRateList {
  items: ReadingRateSummary[];
}

export interface SessionStart {
  session: StudySession;
  progress: ProgressState;
  openUrl: string;
}

export interface SessionResult {
  session: StudySession;
  progress: ProgressState;
}

export interface StartStudySessionInput {
  targetSlug: string;
  lessonId: number;
  materialId: number;
  plannerBlockId?: number;
}

export interface CheckpointStudySessionInput {
  endPage: number;
  elapsedSeconds: number;
  expectedVersion: number;
}

export interface FinishStudySessionInput {
  outcome: Exclude<SessionOutcome, 'skipped'>;
  endPage: number | null;
  elapsedSeconds: number;
  questionsDone: number;
  correctCount: number;
  wrongCount: number;
  doubtCount: number;
  favoriteCount: number;
  notes: string;
  expectedVersion: number;
}

export interface SkipStudySessionInput {
  reason: SkipReason;
  notes: string;
  expectedVersion: number;
}

const progressStatuses = ['unread', 'in_progress', 'covered', 'stale', 'weak', 'strong'] as const;
const sessionStates = ['active', 'finished'] as const;
const sessionOutcomes = ['partial', 'completed', 'failed', 'skipped', 'abandoned'] as const;
const skipReasons = [
  'lack_of_time',
  'fatigue',
  'wrong_material',
  'blocked_prerequisite',
  'too_difficult',
  'other',
] as const;

const isRecord = (value: unknown): value is Record<string, unknown> =>
  Boolean(value) && typeof value === 'object' && !Array.isArray(value);
const isInteger = (value: unknown): value is number =>
  typeof value === 'number' && Number.isInteger(value);
const isPositiveInteger = (value: unknown): value is number =>
  isInteger(value) && value > 0;
const isNonNegativeInteger = (value: unknown): value is number =>
  isInteger(value) && value >= 0;
const isFiniteNumber = (value: unknown): value is number =>
  typeof value === 'number' && Number.isFinite(value);
const isString = (value: unknown): value is string => typeof value === 'string';
const isNullableString = (value: unknown): value is string | null =>
  value === null || isString(value);
const isNullablePositiveInteger = (value: unknown): value is number | null =>
  value === null || isPositiveInteger(value);
const oneOf = <T extends string>(value: unknown, options: readonly T[]): value is T =>
  typeof value === 'string' && options.includes(value as T);

const invalid = (label: string): never => {
  throw new TypeError(`Invalid Study OS ${label} response`);
};

export function parseProgressState(value: unknown): ProgressState {
  if (!isRecord(value)
    || !isPositiveInteger(value.id)
    || !isPositiveInteger(value.lessonId)
    || !isPositiveInteger(value.materialId)
    || !oneOf(value.status, progressStatuses)
    || !isPositiveInteger(value.cursorPage)
    || !isPositiveInteger(value.furthestPage)
    || value.furthestPage < value.cursorPage
    || !isNullableString(value.completedAt)
    || !isNullableString(value.lastSeenAt)
    || !isFiniteNumber(value.confidence)
    || value.confidence < 0
    || value.confidence > 1
    || !isNonNegativeInteger(value.totalSeconds)
    || !isNonNegativeInteger(value.sessionCount)
    || !isPositiveInteger(value.version)) invalid('progress');
  return value as unknown as ProgressState;
}

export function parseStudySession(value: unknown): StudySession {
  if (!isRecord(value)
    || !isPositiveInteger(value.id)
    || !isString(value.idempotencyKey)
    || !value.idempotencyKey.trim()
    || !isString(value.targetSlug)
    || !value.targetSlug.trim()
    || !isPositiveInteger(value.lessonId)
    || !isPositiveInteger(value.materialId)
    || !oneOf(value.state, sessionStates)
    || !isString(value.startedAt)
    || !isNullableString(value.endedAt)
    || !isNonNegativeInteger(value.elapsedSeconds)
    || !isPositiveInteger(value.startPage)
    || !isNullablePositiveInteger(value.endPage)
    || (value.endPage !== null && value.endPage < value.startPage)
    || !isNonNegativeInteger(value.questionsDone)
    || !isNonNegativeInteger(value.correctCount)
    || !isNonNegativeInteger(value.wrongCount)
    || !isNonNegativeInteger(value.doubtCount)
    || !isNonNegativeInteger(value.favoriteCount)
    || !(value.outcome === null || oneOf(value.outcome, sessionOutcomes))
    || !(value.skipReason === null || oneOf(value.skipReason, skipReasons))
    || !isString(value.notes)
    || !isPositiveInteger(value.version)) invalid('session');

  const parsed = value as unknown as StudySession;
  const activeIsValid = parsed.state !== 'active'
    || (parsed.endedAt === null && parsed.outcome === null && parsed.skipReason === null);
  const finishedIsValid = parsed.state !== 'finished'
    || (parsed.endedAt !== null && parsed.outcome !== null);
  const skipIsValid = parsed.outcome === 'skipped'
    ? parsed.skipReason !== null
    : parsed.skipReason === null;
  const pageOutcomeIsValid = !['partial', 'completed'].includes(String(parsed.outcome))
    || parsed.endPage !== null;
  if (!activeIsValid || !finishedIsValid || !skipIsValid || !pageOutcomeIsValid) {
    return invalid('session');
  }
  return parsed;
}

export function parseSessionStart(value: unknown): SessionStart {
  if (!isRecord(value) || !isString(value.openUrl)) return invalid('session start');
  return {
    session: parseStudySession(value.session),
    progress: parseProgressState(value.progress),
    openUrl: value.openUrl,
  };
}

export function parseSessionResult(value: unknown): SessionResult {
  if (!isRecord(value)) return invalid('session result');
  return {
    session: parseStudySession(value.session),
    progress: parseProgressState(value.progress),
  };
}

export function parseMaterialInspection(value: unknown): MaterialInspection {
  if (!isRecord(value)
    || !isPositiveInteger(value.materialId)
    || !isPositiveInteger(value.pageCount)
    || !isNonNegativeInteger(value.pageOffset)) invalid('material inspection');
  return value as unknown as MaterialInspection;
}

function parseReadingRate(value: unknown): ReadingRateSummary {
  if (!isRecord(value)
    || !isPositiveInteger(value.materialId)
    || !isFiniteNumber(value.pagesPerHour)
    || value.pagesPerHour <= 0
    || !isNonNegativeInteger(value.sampleCount)
    || !isNonNegativeInteger(value.totalSeconds)
    || !oneOf(value.source, ['default', 'observed'] as const)) invalid('reading rate');
  return value as unknown as ReadingRateSummary;
}

export function parseReadingRateList(value: unknown): ReadingRateList {
  if (!isRecord(value) || !Array.isArray(value.items)) return invalid('reading rate list');
  return { items: value.items.map(parseReadingRate) };
}

const jsonHeaders = { 'Content-Type': 'application/json' };

const materialQuery = (targetSlug: string, lessonId: number, materialId: number) =>
  new URLSearchParams({
    targetSlug,
    lessonId: String(lessonId),
    materialId: String(materialId),
  });

export async function fetchProgress(
  targetSlug: string,
  lessonId: number,
  materialId: number,
  signal?: AbortSignal,
): Promise<ProgressState> {
  const query = materialQuery(targetSlug, lessonId, materialId);
  return parseProgressState(await requestJson(`/api/v1/progress?${query}`, { signal }));
}

export async function inspectMaterial(
  materialId: number,
  targetSlug: string,
): Promise<MaterialInspection> {
  const query = new URLSearchParams({ targetSlug });
  return parseMaterialInspection(await requestJson(
    `/api/v1/materials/${materialId}/inspect?${query}`,
    { method: 'POST' },
  ));
}

export async function fetchReadingRates(
  targetSlug: string,
  signal?: AbortSignal,
): Promise<ReadingRateList> {
  const query = new URLSearchParams({ targetSlug });
  return parseReadingRateList(await requestJson(`/api/v1/reading-rates?${query}`, { signal }));
}

export async function startStudySession(
  input: StartStudySessionInput,
  idempotencyKey: string,
): Promise<SessionStart> {
  return parseSessionStart(await requestJson('/api/v1/sessions', {
    method: 'POST',
    headers: { ...jsonHeaders, 'Idempotency-Key': idempotencyKey },
    body: JSON.stringify(input),
  }));
}

export async function fetchActiveStudySession(
  targetSlug: string,
  lessonId: number,
  materialId: number,
  signal?: AbortSignal,
): Promise<StudySession> {
  const query = materialQuery(targetSlug, lessonId, materialId);
  return parseStudySession(await requestJson(`/api/v1/sessions/active?${query}`, { signal }));
}

export async function fetchStudySession(
  sessionId: number,
  targetSlug: string,
  signal?: AbortSignal,
): Promise<StudySession> {
  const query = new URLSearchParams({ targetSlug });
  return parseStudySession(await requestJson(`/api/v1/sessions/${sessionId}?${query}`, { signal }));
}

export async function checkpointStudySession(
  sessionId: number,
  input: CheckpointStudySessionInput,
): Promise<SessionResult> {
  return parseSessionResult(await requestJson(`/api/v1/sessions/${sessionId}`, {
    method: 'PATCH',
    headers: jsonHeaders,
    body: JSON.stringify(input),
  }));
}

export async function finishStudySession(
  sessionId: number,
  input: FinishStudySessionInput,
): Promise<SessionResult> {
  return parseSessionResult(await requestJson(`/api/v1/sessions/${sessionId}/finish`, {
    method: 'POST',
    headers: jsonHeaders,
    body: JSON.stringify(input),
  }));
}

export async function skipStudySession(
  sessionId: number,
  input: SkipStudySessionInput,
): Promise<SessionResult> {
  return parseSessionResult(await requestJson(`/api/v1/sessions/${sessionId}/skip`, {
    method: 'POST',
    headers: jsonHeaders,
    body: JSON.stringify(input),
  }));
}
